### PR TITLE
#2mtq0yn #statemine transfers

### PR DESCRIPTION
**Pull request by dzmitry-lahoda**
- simulation of registering and trasfering statemine assets in both direction like in acala
- all needed assets and xcmp setups for that (decimals, fees)

### CODE REVIEW

File: Makefile.toml [hunk 1]
```diff
@@ -60,7 +60,7 @@ workspace = false
 [tasks.run-local-integration-tests-debug]
 description = "fastest way to build and debug runtime in simulator"
 script = '''
-RUST_BACKTRACE=full RUST_LOG=trace,parity-db=warn,trie=warn,runtime=trace,substrate-relay=trace,bridge=trace,xcmp=trace cargo +stable test  --package local-integration-tests --features local-integration-tests --features picasso  --no-default-features -- --nocapture --test-threads=1
+RUST_BACKTRACE=full RUST_LOG=trace,parity-db=warn,trie=warn,runtime=trace,substrate-relay=trace,bridge=trace,xcmp=trace,xcm=trace cargo +stable test  --package local-integration-tests --features local-integration-tests --features picasso  --no-default-features -- --nocapture --test-threads=1
```
Comment (KaiserKarel) [inline Makefile.toml:63]: This is quite a long command, perhaps split this using the `env` key for the task?

**Comment by dzmitry-lahoda** [comment]
running via special command with thead = 1 passes. so running as part of general tests it seems fails
```
failures:

---- cross_chain_transfer::transfer_from_relay_chain_deposit_to_treasury_if_below_existential_deposit stdout ----
thread 'cross_chain_transfer::transfer_from_relay_chain_deposit_to_treasury_if_below_existential_deposit' panicked at 'assertion failed: `(left == right)`
  left: `9935999999`,
 right: `0`: assets did not get to recipient as it is not enough to pay ED', integration-tests/local-integration-tests/src/cross_chain_transfer.rs:384:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    cross_chain_transfer::transfer_from_relay_chain_deposit_to_treasury_if_below_existential_deposit

test result: FAILED. 28 passed; 1 failed; 1 ignored; 0 measured; 0 filtered out; finished in 0.14s
```
running CI test fully locally checking why is that case

File: integration-tests/local-integration-tests/src/statemine.rs [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+// transfer custom asset from this chain  to Statemine
+fn this_chain_side(fee_amount: u128, foreign_asset_id_on_this: CurrencyId) {
+	This::execute_with(|| {
+		use this_runtime::*;
+
+		let bob_statemine_asset_amount =
+			Tokens::free_balance(foreign_asset_id_on_this, &AccountId::from(BOB));
+		assert_eq!(9_999_936_000_000, bob_statemine_asset_amount,);
```
Comment (vimukthi-git) [inline integration-tests/local-integration-tests/src/statemine.rs:128]: If I understand correctly this is `TEN` - fee ?

**Review by vimukthi-git — APPROVED** [review]
Helpful integration tests :)

File: frame/assets-registry/src/runtime.rs [hunk 3]
```diff
@@ -96,6 +91,10 @@ impl pallet_assets_registry::Config for Runtime {
 		EnsureSignedBy<RootAccount, AccountId>, // for tests
 		EnsureRoot<AccountId>,                  // for benchmarks
 	>;
+	type ParachainOrGovernanceOrigin = EnsureOneOf<
+		EnsureSignedBy<RootAccount, AccountId>, // for tests
+		EnsureRoot<AccountId>,                  // for benchmarks
```
[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply frame/assets-registry/src/runtime.rs:96]: it was here originally, i do not know what is for, did not changed. but as i get it the way to make root origin in any run. so will not change it.

File: integration-tests/local-integration-tests/src/assets_integration.rs [hunk 4]
```diff
@@ -9,27 +9,50 @@ use crate::{
...
+			RawOrigin::Root.into(),
+			CurrencyId(123),
+			XcmAssetLocation(MultiLocation::new(1, X1(Parachain(4321)))),
+			Ratio::checked_from_rational(10u32, 100u32),
+			None,
+		)
+		.unwrap();
+		assert_eq!(
+			1000,
+			<PriceConverter<AssetsRegistry>>::get_price_inverse(CurrencyId(42), 100).unwrap()
```
[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply integration-tests/local-integration-tests/src/assets_integration.rs:48]: yes)))

File: integration-tests/local-integration-tests/src/statemine.rs (hunk 2 above)
Comment (dzmitry-lahoda) [reply integration-tests/local-integration-tests/src/statemine.rs:128]: Maybe yes. so it is strait port from Acala (they sometimes has calculation and sometimes hardcode), will improve later (need to get exact fee values from other runtimes)

**Review by vimukthi-git — APPROVED** [review]
LGTM

File: frame/assets-registry/src/lib.rs [hunk 5]
```diff
@@ -173,6 +189,29 @@ pub mod pallet {
...
+			origin: OriginFor<T>,
+			target_parachain_id: ParaId,
+			asset_id: T::ForeignAssetId,
+			minimal_amount: Option<T::Balance>,
+		) -> DispatchResultWithPostInfo {
+			T::ParachainOrGovernanceOrigin::ensure_origin(origin)?;
+			// TODO: in case it is set to parachain, check that chain can target only its origin
+			MinFeeAmounts::<T>::mutate_exists(target_parachain_id, asset_id, |x| {
+				*x = minimal_amount
+			});
```
Comment (andor0) [inline frame/assets-registry/src/lib.rs:212]: I would add a event about success.This can be useful for UI and tests.

Comment (dzmitry-lahoda) [reply frame/assets-registry/src/lib.rs:212]: @andor0 will add in next PR

File: runtime/common/src/lib.rs [hunk 6]
```diff
@@ -153,7 +163,24 @@ impl MinimalOracle for PriceConverter {
...
+							))
+						}
+					} else {
+						Err(DispatchError::Other(
+							xcmp_errors::AMOUNT_OF_ASSET_IS_MORE_THAN_MAX_POSSIBLE,
+						))
+					}
+				} else {
+					Err(DispatchError::Other(xcmp_errors::ASSET_IS_NOT_PRICEABLE))
+				},
```
Comment (andor0) [inline runtime/common/src/lib.rs:183]:
```
_ => {
    let ratio = match AssetsRegistry::get_ratio(asset_id) {
        Some(ratio) => ratio,
        None => return Err(DispatchError::Other(xcmp_errors::ASSET_IS_NOT_PRICEABLE)),
    };
    let amount = match Ratio::checked_from_integer(amount) {
        Some(amount) => amount,
        None => return Err(DispatchError::Other(
            xcmp_errors::AMOUNT_OF_ASSET_IS_MORE_THAN_MAX_POSSIBLE,
        )),
    };
    match ratio.checked_mul(&amount) {
        Some(payment) => payment.into_inner().safe_div(&Ratio::accuracy()).map_err(Into::into),
        None => return Err(DispatchError::Other(
            xcmp_errors::AMOUNT_OF_ASSET_IS_MORE_THAN_MAX_POSSIBLE,
        )),
    }
}
```
?

**Review by andor0 — APPROVED** [review]

File: runtime/common/src/lib.rs (hunk 6 above)
Comment (dzmitry-lahoda) [reply runtime/common/src/lib.rs:183]: will fix in next pr via and_then

**Pull request merged**